nevergrad/functions/powersystems/core.py

```diff
@@ -0,0 +1,234 @@
+            ) -> None:
+        # Number of stocks (dams).
+        self.N = num_stocks
+        N = self.N
+        # Parameters describing the problem.
+        self.year_to_day_ratio = year_to_day_ratio
+        self.constant_to_year_ratio = constant_to_year_ratio
+        self.back_to_normal = back_to_normal
+        self.consumption_noise = consumption_noise
+        self.num_thermal_plants = num_thermal_plants
+        self.number_of_years = num_years 
+        self.failure_cost = failure_cost
+        
+        self.average_consumption = self.constant_to_year_ratio * self.year_to_day_ratio
+        self.thermal_power_capacity = self.average_consumption * np.random.rand(self.num_thermal_plants)
+        self.thermal_power_prices = np.random.rand(num_thermal_plants)
+        dam_managers: List[Any] = []
+        for i in range(N):
+            dam_managers += [Agent(10 + N + 2*self.num_thermal_plants, depth, width)]
+        the_dimension = sum([a.GetParamNumbers() for a in dam_managers])
+        self.dam_managers = dam_managers
+        super().__init__(self._simulate_power_system, Instrumentation(inst.var.Array(the_dimension)))
+        self._descriptors.update(num_stocks=num_stocks, depth=depth, width=width)
+
+    def get_num_vars(self) -> List[Any]:
+        return [m.GetParamNumbers() for m in self.dam_managers]
+
+    def _simulate_power_system(self, x: np.ndarray) -> float:
+        failure_cost = self.failure_cost  # Cost of power demand which is not satisfied (equivalent to a expensive infinite thermal group).
+        dam_managers = self.dam_managers
+        for a in dam_managers:
+            assert(len(x) >= a.GetParamNumbers())
+            a.SetParams(np.array(x[:a.GetParamNumbers()]))
+            x = x[a.GetParamNumbers():]
+        assert(len(x) == 0)
+        N = int(self.N)
+        # Assume empty initial stocks.
+        stocks = [0.] * N   
+        # Nonsense delays.
+        delay = [math.cos(i) for i in range(N)]
+        cost = 0.
+        # Loop on time steps.
+        num_time_steps = int(365*24*self.number_of_years)
+        consumption = 0.
+        hydro_prod_per_time_step: List[Any] = []
+        consumption_per_time_step: List[float] = []
+        for t in range(num_time_steps):
+    
+            # Rain
+            for i in range(N):
+                stocks[i] += 0.5*(1.+math.cos(2*pi*t/(24*365) + delay[i])) * np.random.rand()
+            # Consumption model.
+            base_consumption = (self.constant_to_year_ratio*self.year_to_day_ratio 
+                    +0.5*self.year_to_day_ratio*(1.+math.cos(2*pi*t/(24*365))) + 0.5*(1.+math.cos(2*pi*t/24)))
+            if t == 0:
+                consumption = base_consumption
+            else:
+                consumption = max(0., consumption + self.consumption_noise*(np.random.rand()-.5) + self.back_to_normal * (base_consumption - consumption))
+            consumption_per_time_step += [consumption]
+            # Water usage.
```

Comment:
Shouldn't the comment say energy usage ? Since you can also use thermal power plants.

Reply:
Fixed.

Reply:
Yes, fixed.